client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -18,55 +19,81 @@ import FormLabel from 'components/forms/form-label';
 import InfoPopover from 'components/info-popover';
 import { addQueryArgs } from 'lib/url';
 
-function SiteIconSetting( { translate, siteId, isJetpack, customizerUrl, generalOptionsUrl } ) {
-	let buttonProps;
-	if ( isEnabled( 'manage/site-settings/site-icon' ) ) {
-		buttonProps = {
-			type: 'button'
-		};
-	} else {
-		buttonProps = { rel: 'external' };
+class SiteIconSetting extends Component {
+	static propTypes = {
+		translate: PropTypes.func,
+		siteId: PropTypes.number,
+		isJetpack: PropTypes.bool,
+		customizerUrl: PropTypes.string,
+		generalOptionsUrl: PropTypes.string
+	};
+
+	state = {
+		isModalVisible: false
+	};
+
+	toggleModal = ( isModalVisible ) => this.setState( { isModalVisible } );
 
-		if ( isJetpack ) {
-			buttonProps.href = addQueryArgs( {
-				'autofocus[section]': 'title_tagline'
-			}, customizerUrl );
+	hideModal = () => this.toggleModal( false );
+
+	showModal = () => this.toggleModal( true );
+
+	preloadModal() {
+		asyncRequire( 'post-editor/media-modal' );
+	}
+
+	render() {
+		const { translate, siteId, isJetpack, customizerUrl, generalOptionsUrl } = this.props;
+		const isIconManagementEnabled = isEnabled( 'manage/site-settings/site-icon' );
+
+		let buttonProps;
+		if ( isIconManagementEnabled ) {
+			buttonProps = {
+				type: 'button',
+				onClick: this.showModal,
+				onMouseEnter: this.preloadModal
+			};
 		} else {
-			buttonProps.href = generalOptionsUrl;
-			buttonProps.target = '_blank';
+			buttonProps = { rel: 'external' };
```

Comment:
We should probably add an external icon to the button when it matches this case.